### ISSUE TITLE
Add `snapshot_user` role to `editor` role

### ISSUE THREAD

**Issue by kunisen** [elastic/elasticsearch#103247]
## Background

Currently, it seems ESS editor role doesn’t allow taking snapshot, which is not ideal.
https://www.elastic.co/guide/en/cloud/current/ec-user-privileges.html

![image](https://github.com/elastic/elasticsearch/assets/30574753/9ca4bf0e-2d69-48e8-a9b1-9ab2427bf535)

Got this error when click `Take snapshot now` button:

```json
{
    "status": 403, 
    "error": {
        "root_cause": [
            {
                "reason": "action [cluster:admin/slm/execute] is unauthorized for user [elastic-editor-proxy] with effective roles [editor], this action is granted by the cluster privileges [manage_slm,manage,all]", 
                "type": "security_exception"
            }
        ], 
        "type": "security_exception", 
        "reason": "action [cluster:admin/slm/execute] is unauthorized for user [elastic-editor-proxy] with effective roles [editor], this action is granted by the cluster privileges [manage_slm,manage,all]"
    }
}
```

　
While digging around, I found this seems to be by design, as ESS `editor` maps to Stack `editor`
https://www.elastic.co/guide/en/cloud/current/ec-stack-user-org-member.html

And stack editor doesn’t seem to allow taking snapshot
https://www.elastic.co/guide/en/elasticsearch/reference/8.11/built-in-roles.html#built-in-roles
```
Grants full access to all features in Kibana (including Solutions) and read-only access to data indices.
```

Also, it seems not easy to workaround the issue. The ESS role editor maps to stack editor role, but the stack editor role is reserved, and not able to change or add extra permission piece.

Which it would be really complex to workaround, as we need to 
- either use ESS `admin` role (too wide) 
- or use stack basic auth to login (can’t use the email to SSO login).

## Feature request

Can we add `snapshot_user` to `editor` role, so that we can let deployment `editor` be able to take snapshot, while not having access to indices directly?
Thanks!

**Comment by bytebilly** [comment]
As you mentioned, Editors are not supposed to manage snapshot or any other "administrative" task. This is intentional to keep expectations clear when using predefined roles. Adding the ability to create snapshots would open up discussions for any other administrative feature, breaking the "simple rule" that we have today.

If customers want to change the behaviour, they can create an additional role mapping at the deployment level so that every user with `editor` also has `snapshot_user`. The limitation with this approach is that it must be done for each deployment individually.

In the future, we may introduce improvements to customize ESS roles and eventually add more flexibility.

@alxchalkias @Aviad-Abutbul We should probably modify the Cloud UI and remove the snapshot actions if they are not granted to Editors, and possibly improve our docs to explicitly mention that.

**Comment by alxchalkias** [comment]
@bytebilly you raise a good point, but my understanding is the opposite of yours. We'd discussed with @Kushmaro in the past that the main difference between Admin and Editor is the security controls only granted to the former. You can see this documented in ESS today ([link](https://www.elastic.co/guide/en/cloud/current/ec-user-privileges.html)). We shouldn't get into the customization of roles topic here, but if there's no major concern on the stack side, I think we should accommodate Kuni's request to modify `editor` to include `snapshot_user`

**Comment by bytebilly** [comment]
> but if there's no major concern on the stack side

I actually have, for the reasons listed in my previous message.
Unless we want to revisit the entire concept of Editor, but in this case we should add any other non-security related feature, like managing snapshots, templates, ILM, pipelines, transforms, etc.
That would be a huge shift in the scope of Editor, and it's hard to not consider it a breaking change subject to our internal policy, so that it would not be effective for a very long time anyway.

This is the way Editor has been conceived in agreement with Solutions, and even if I understand the reasons behind this request, I don't think that they are strong enough to force this foundational change, also given that there is an alternative way of doing that (even if it requires some additional configuration by the customer).

The ESS documentation looks very specific to the difference of these roles wrt Cloud actions, but it's very far from being true at the Stack level, where none of the administrative tasks granted to Admins are granted to Editors. The other main difference between the two is that Editor cannot ingest data, while Admin can. Maybe we can make it more explicit by saying that "Editor has the same rights as Admin **in Elastic Cloud**".

**Comment by Kushmaro** [comment]
@bytebilly , indeed in Elastic Cloud there's not much difference (intentionally, but also, due to reasons of very limited scope of capabilities for both roles) between Editor and Admin.. So generally, i'd also be +1 to enable the Editor to "edit a deployments data via snapshots" : ) 

 I'm not sure why in the stack we opted to not allow editor to edit data on stack indices? that seems counter-intuitive for the meaning of the word "editor"

**Comment by bytebilly** [comment]
> I'm not sure why in the stack we opted to not allow editor to edit data on stack indices? that seems counter-intuitive for the meaning of the word "editor"

These predefined roles are primarily intended for Solutions scenarios, and Editor has been tuned to configure the solution, not to ingest data, that is generally not done by an interactive user in these cases.

> Also related to this, I'm not sure if this is expected but the "editor" can actually restore a snapshot from the UI by clicking the Restore from another deployment button (besides not being able to take one). It's quite confusing.

This is actually not expected. Is the restore process successful, or just allowed in the UI?

**Comment by Kushmaro** [comment]
>These predefined roles are primarily intended for Solutions scenarios, and Editor has been tuned to configure the solution, not to ingest data, that is generally not done by an interactive user in these cases.

I think this "tuning" will cause quite confusion for folks who are not familiar with the stack.. the name "editor" implies being able to "edit"

and that's also the users' first encounter with the role when they enter the cloud console.

Also its worthwhile Ingesting data (creating new documents) isn't the same as editing existing data.

**Comment by renshuki** [comment]
> This is actually not expected. Is the restore process successful, or just allowed in the UI?

Yes, it triggers a plan change and successfully restore the data.
